File: interpreter.scala
case class Context(context: Map[String, Int]) {
  def get(name: String): Int = context(name)
}

abstract class Expression {
  def interpret(ctx: Context): Int
}

case class Digit(value: Int) extends Expression {
  override def interpret(ctx: Context): Int = value
}

case class Variable(name: String) extends Expression {
  override def interpret(ctx: Context): Int = ctx.get(name)
}

case class Plus(lh: Expression, rh: Expression) extends Expression {
  override def interpret(ctx: Context): Int = lh.interpret(ctx) + rh.interpret(ctx)
}

case class Times(lh: Expression, rh: Expression) extends Expression {
  override def interpret(ctx: Context): Int = lh.interpret(ctx) * rh.interpret(ctx)
}

object Interpreter extends Application {
  val exp = Plus ( Digit(9), Times ( Variable("a"), Digit(4) ) )
  val ctx = Context(Map("a" -> 3))
  println("a 4 * 9 + = " + exp.interpret(ctx))
}

File: interpreter.hs
import Data.Map (Map)
import qualified Data.Map as M

data Expression a = Plus   (Expression a) (Expression a)
                  | Minus  (Expression a) (Expression a)
                  | Times  (Expression a) (Expression a)
                  | Divide (Expression a) (Expression a)
                  | Variable String
                  | Number a

class Interpreter i where
    interpret :: Fractional a => i a -> Map String a -> a

instance Interpreter Expression where
    interpret (Plus a b) m = interpret a m + interpret b m
    interpret (Minus a b) m = interpret a m - interpret b m
    interpret (Times a b) m = interpret a m * interpret b m
    interpret (Divide a b) m = interpret a m / interpret b m
    interpret (Variable name) m = case M.lookup name m of
                                       Nothing -> error $ "Variable " ++ name ++ " not found"
                                       Just x -> x
    interpret (Number a) _ = a

test :: Fractional a => Expression a
test = Plus
        (Times
            (Variable "x")
            (Variable "y"))
        (Minus
            (Number 1)
            (Number 3.5))

variables :: Map String Double
variables = M.fromList [("x", 5), ("y", 6)]

main :: IO ()
main = print $ interpret test variables

